File: Stage24/C/Problem1629.c
#include <stdio.h>

long long A, B, C;

long long power(long long a, long long b) {
    if (b == 1)
        return a % C;

    long long half = power(a, b / 2);

    if (b % 2 == 0)
        return (half * half) % C;
    else
        return (half * half % C * a) % C;
}

int main(void) {
    scanf("%lld %lld %lld", &A, &B, &C);
    printf("%lld\n", power(A, B));
    return 0;
}

File: Stage24/Python/Problem1629.py
import sys
input = sys.stdin.readline

def power(a, b, c):
    if b == 1:
        return a % c

    half = power(a, b // 2, c)

    if b % 2 == 0:
        return (half * half) % c
    else:
        return (half * half * a) % c

def main():
    A, B, C = map(int, input().split())
    print(power(A, B, C))

if __name__ == "__main__":
    main()